lib/grandpa/finalization_test.go

```diff
@@ -0,0 +1,260 @@
+						timeToFail := time.NewTimer(failTime)
+						<-timeToFail.C
+						return errors.New("mocked finalisation engine fails")
+					})
+					mockEngine.EXPECT().Stop().Return(nil)
+
+					// once the finalisation engine fails the finalisation handler
+					// should be awere of the error and call the stop method from the
+					// voting round which will release the start method from voting round service
+					votingStopChannel := make(chan struct{})
+					mockVoting := NewMockephemeralService(ctrl)
+					mockVoting.EXPECT().Start().DoAndReturn(func() error {
+						select {
+						case <-time.After(failTime + time.Second):
+							return errors.New("timeout waiting votingStopChannel")
+						case <-votingStopChannel:
+							return nil
+						}
+					})
+					mockVoting.EXPECT().Stop().DoAndReturn(func() error {
+						close(votingStopChannel)
+						return nil
+					})
+
+					return mockEngine, mockVoting
+				}
+
+				return &finalizationHandler{
+					newServices:    builder,
+					timeoutStop:    2 * time.Second,
+					observableErrs: make(chan error),
+					stopCh:         make(chan struct{}),
+					handlerDone:    make(chan struct{}),
+				}
+			},
+		},
+	}
+
+	for tname, tt := range tests {
+		tt := tt
+
+		t.Run(tname, func(t *testing.T) {
+			t.Parallel()
+
+			ctrl := gomock.NewController(t)
+			finalizationHandler := tt.createFinalizationHandler(ctrl)
+
+			err := finalizationHandler.waitServices()
+			if tt.wantErr != nil {
+				require.Error(t, err)
+				require.EqualError(t, err, tt.wantErr.Error())
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
+func Test_FinalizationHandler_Stop_ShouldHalt_Services(t *testing.T) {
+	t.Parallel()
+
+	ctrl := gomock.NewController(t)
+	builder := func() (engine ephemeralService, voting ephemeralService) {
+		engineStopCh := make(chan struct{})
+		mockEngine := NewMockephemeralService(ctrl)
+		mockEngine.EXPECT().Start().DoAndReturn(func() error {
+			<-engineStopCh
+			return nil
+		})
+		mockEngine.EXPECT().Stop().DoAndReturn(func() error {
+			close(engineStopCh)
+			return nil
+		})
+
+		votingStopCh := make(chan struct{})
+		mockVoting := NewMockephemeralService(ctrl)
+		mockVoting.EXPECT().Start().DoAndReturn(func() error {
+			<-votingStopCh
+			return nil
+		})
+		mockVoting.EXPECT().Stop().DoAndReturn(func() error {
+			close(votingStopCh)
+			return nil
+		})
+
+		return mockEngine, mockVoting
+	}
+
+	handler := &finalizationHandler{
+		newServices: builder,
+		// mocked initiate round function
+		initiateRound:  func() error { return nil },
+		timeoutStop:    2 * time.Second,
+		observableErrs: make(chan error),
+		stopCh:         make(chan struct{}),
+		handlerDone:    make(chan struct{}),
+	}
+
+	doneCh := make(chan struct{})
+	errsCh, err := handler.Start()
+	require.NoError(t, err)
+
+	go func() {
+		defer close(doneCh)
+		for err := range errsCh {
+			t.Errorf("expected no error, got %s", err)
+			return
+		}
```

Comment:
this looks sketchy... I think this just quickly goes over the empty error channel and exits; shouldn't we check for errors later in a blocking manner with i.e. a `select+default`?